### PR TITLE
Add support for starting values

### PR DESCRIPTION
**Pull request by blegat**
@ccoffrin Can you test it and report whether it works as expected for your use case ?

Closes https://github.com/JuliaOpt/SCS.jl/issues/109

### CODE REVIEW

**Comment by ccoffrin** [comment]
Using JuMP master, MOI v0.6.1 and this branch I am getting,

```
ERROR: MethodError: no method matching allocate(::SCS.Optimizer, ::MathOptInterface.ObjectiveFunction{MathOptInterface.ScalarQuadraticFunction{Float64}}, ::MathOptInterface.ScalarQuadraticFunction{Float64})
Closest candidates are:
  allocate(::MathOptInterface.Utilities.AbstractModel, ::Any...) at /home/.julia/v0.6/MathOptInterface/src/Utilities/model.jl:339
  allocate(::MathOptInterface.Utilities.MockOptimizer, ::MathOptInterface.ObjectiveFunction, ::Any) at /home/.julia/v0.6/MathOptInterface/src/Utilities/mockoptimizer.jl:285
  allocate(::MathOptInterface.Utilities.MockOptimizer, ::Union{MathOptInterface.AbstractConstraintAttribute, MathOptInterface.AbstractModelAttribute, MathOptInterface.AbstractOptimizerAttribute, MathOptInterface.AbstractVariableAttribute}, ::Any) at /home/.julia/v0.6/MathOptInterface/src/Utilities/mockoptimizer.jl:284
  ...
Stacktrace:
 [1] _pass_attributes(::SCS.Optimizer, ::MathOptInterface.Utilities.UniversalFallback{JuMP.JuMPMOIModel{Float64}}, ::Bool, ::MathOptInterface.Utilities.IndexMap, ::Array{MathOptInterface.AbstractModelAttribute,1}, ::Tuple{}, ::Tuple{}, ::Tuple{}, ::MathOptInterface.Utilities.#allocate) at /home/.julia/v0.6/MathOptInterface/src/Utilities/copy.jl:59
 [2] allocate_load(::SCS.Optimizer, ::MathOptInterface.Utilities.UniversalFallback{JuMP.JuMPMOIModel{Float64}}, ::Bool) at /home/.julia/v0.6/MathOptInterface/src/Utilities/copy.jl:248
 [3] (::MathOptInterface.#kw##copy_to)(::Array{Any,1}, ::MathOptInterface.#copy_to, ::SCS.Optimizer, ::MathOptInterface.Utilities.UniversalFallback{JuMP.JuMPMOIModel{Float64}}) at ./<missing>:0
 [4] attachoptimizer!(::MathOptInterface.Utilities.CachingOptimizer{MathOptInterface.AbstractOptimizer,MathOptInterface.Utilities.UniversalFallback{JuMP.JuMPMOIModel{Float64}}}) at /home/.julia/v0.6/MathOptInterface/src/Utilities/cachingoptimizer.jl:125
 [5] optimize!(::MathOptInterface.Utilities.CachingOptimizer{MathOptInterface.AbstractOptimizer,MathOptInterface.Utilities.UniversalFallback{JuMP.JuMPMOIModel{Float64}}}) at /home/.julia/v0.6/MathOptInterface/src/Utilities/cachingoptimizer.jl:158
 [6] optimize!(::MathOptInterface.Bridges.LazyBridgeOptimizer{MathOptInterface.Utilities.CachingOptimizer{MathOptInterface.AbstractOptimizer,MathOptInterface.Utilities.UniversalFallback{JuMP.JuMPMOIModel{Float64}}},MathOptInterface.Bridges.AllBridgedConstraints{Float64}}) at /home/.julia/v0.6/MathOptInterface/src/Bridges/bridgeoptimizer.jl:73
 [7] #optimize!#92(::Bool, ::Function, ::JuMP.Model, ::Void) at /home/.julia/v0.6/JuMP/src/optimizer_interface.jl:65
 [8] macro expansion at ./util.jl:378 [inlined]
 [9] optimize!(::PowerModels.GenericPowerModel{PowerModels.SDPWRMForm}) at /home/.julia/v0.6/PowerModels/src/core/base.jl:126
 [10] #solve_generic_model#83(::Function, ::Function, ::PowerModels.GenericPowerModel{PowerModels.SDPWRMForm}, ::SCS.Optimizer) at /home/.julia/v0.6/PowerModels/src/core/base.jl:201
 [11] (::PowerModels.#kw##solve_generic_model)(::Array{Any,1}, ::PowerModels.#solve_generic_model, ::PowerModels.GenericPowerModel{PowerModels.SDPWRMForm}, ::SCS.Optimizer) at ./<missing>:0
 [12] #run_generic_model#80(::Function, ::Array{Any,1}, ::Function, ::Dict{String,Any}, ::Type{T} where T, ::SCS.Optimizer, ::Function) at /home/.julia/v0.6/PowerModels/src/core/base.jl:171
 [13] run_generic_model(::Dict{String,Any}, ::Type{T} where T, ::SCS.Optimizer, ::Function) at /home/.julia/v0.6/PowerModels/src/core/base.jl:169
 [14] #run_generic_model#79(::Array{Any,1}, ::Function, ::String, ::Type{T} where T, ::SCS.Optimizer, ::Function) at /home/.julia/v0.6/PowerModels/src/core/base.jl:164
 [15] #run_opf#325 at /home/.julia/v0.6/PowerModels/src/prob/opf.jl:15 [inlined]
 [16] run_opf(::String, ::Type{T} where T, ::SCS.Optimizer) at /home/.julia/v0.6/PowerModels/src/prob/opf.jl:15
```

File: src/MOIWrapper.jl [hunk 1]
```diff
@@ -259,13 +260,46 @@ function MOIU.load_variables(optimizer::Optimizer, nvars::Integer)
     b = zeros(m)
     c = zeros(nvars)
     optimizer.data = ModelData(m, nvars, I, J, V, b, 0., c)
+    if length(optimizer.sol.primal) != nvars
```
Comment (mlubin) [inline src/MOIWrapper.jl]: What could the length be if not `nvars`? Would `isempty` also work as a test? Same below.

**Comment by mlubin** [comment]
Do we have a test for this functionality anywhere?

**Comment by blegat** [comment]
Yes, it is tested:
https://github.com/JuliaOpt/JuMP.jl/blob/500945cac4949dfd9e30cb0ca68450f19f4ff34d/test/objective.jl#L24
and
https://github.com/JuliaOpt/JuMP.jl/blob/500945cac4949dfd9e30cb0ca68450f19f4ff34d/src/aff_expr.jl#L317

**Comment by ccoffrin** [comment]
Do you have a test were the model has conic constraints and a linear objective?

**Comment by ccoffrin** [comment]
In any case, this PR seems to have resolved #109, should we move this later point to a separate thread?

**Comment by blegat** [comment]
Yes we can open an issue on JuMP, if you could produce a MWE producing a quadratic objective when it should be linear, it would help. It is probably because of a promotion that shouldn't happen.
For instance if you do
```julia
@objective(model, [1, 2]' * [x, y])
```
if the product is handled by a function that start by promoting the two vectors to the same type, it will promote them to affine expressions and hence the result would be quadratic.
I would guess that you have somthing similar.

File: src/MOIWrapper.jl (hunk 1 above)
Comment (blegat) [reply src/MOIWrapper.jl]: `sol` contains the solution of the previous optimization. If the user has added new variables, then it is not `nvars`

Comment (mlubin) [reply src/MOIWrapper.jl]: Does this interface allow adding variables and/or constraints?

Comment (blegat) [reply src/MOIWrapper.jl]: You can copy a model with a different number of variables/constraints. Normally, they will arrive in the same order.

Comment (mlubin) [reply src/MOIWrapper.jl]: Could this `if` be replaced with `resize!` then?

**Comment by ccoffrin** [comment]
@blegat good news, I found a bug in my code, which was the cause of the quadratic objective function.  That is fixed now.

Now I am getting this error,
```
ERROR: MethodError: no method matching _allocate_constraint(::SCS.ConeData, ::MathOptInterface.VectorAffineFunction{Float64}, ::MathOptInterface.PositiveSemidefiniteConeSquare)
Closest candidates are:
  _allocate_constraint(::SCS.ConeData, ::Any, ::Union{MathOptInterface.EqualTo, MathOptInterface.Zeros}) at home/.julia/v0.6/SCS/src/MOIWrapper.jl:100
  _allocate_constraint(::SCS.ConeData, ::Any, ::Union{MathOptInterface.GreaterThan, MathOptInterface.LessThan, MathOptInterface.Nonnegatives, MathOptInterface.Nonpositives}) at home/.julia/v0.6/SCS/src/MOIWrapper.jl:106
  _allocate_constraint(::SCS.ConeData, ::Any, ::MathOptInterface.SecondOrderCone) at home/.julia/v0.6/SCS/src/MOIWrapper.jl:112
  ...
Stacktrace:
 [1] allocate_constraint at home/.julia/v0.6/SCS/src/MOIWrapper.jl:132 [inlined]
 [2] allocate_constraints(::SCS.Optimizer, ::MathOptInterface.Utilities.UniversalFallback{JuMP.JuMPMOIModel{Float64}}, ::Bool, ::MathOptInterface.Utilities.IndexMap, ::Type{MathOptInterface.VectorAffineFunction{Float64}}, ::Type{MathOptInterface.PositiveSemidefiniteConeSquare}) at home/.julia/v0.6/MathOptInterface/src/Utilities/copy.jl:205
 [3] allocate_load(::SCS.Optimizer, ::MathOptInterface.Utilities.UniversalFallback{JuMP.JuMPMOIModel{Float64}}, ::Bool) at home/.julia/v0.6/MathOptInterface/src/Utilities/copy.jl:253
```
Recommendations?

**Comment by ccoffrin** [comment]
This works with SCS in JuMP v0.18 / MPB,
```
    WR = @variable(m, [1:n, 1:n], Symmetric)
    WI = @variable(m, [1:n, 1:n])
   @SDconstraint(pm.model, [WR WI; -WI WR] >= 0)
```
What would be the translation of this into JuMP v0.19 / MOI?

**Comment by blegat** [comment]
> What would be the translation of this into JuMP v0.19 / MOI?

Same thing.

I just realized that you might get the error you give in https://github.com/JuliaOpt/SCS.jl/pull/111#issuecomment-425598620 if you do not provide the solver in the `Model` constructor but rather in the `optimize!` call. Is that the case ?

**Comment by ccoffrin** [comment]
To finish up this discussion.  You are correct, specifying the optimizer in the JuMP model constructor resolved the most recent issue.  The only remaining issue I have at the moment is with setting the solver parameters (i.e. https://github.com/JuliaOpt/SCS.jl/issues/113).  In any case, this starting point fix is working great for me.

**Pull request merged**